Lib/ntpath.py

```diff
@@ -641,22 +639,65 @@ def commonpath(paths):
         raise
 
 
-# determine if two files are in fact the same file
 try:
-    # GetFinalPathNameByHandle is available starting with Windows 6.0.
-    # Windows XP and non-Windows OS'es will mock _getfinalpathname.
-    if sys.getwindowsversion()[:2] >= (6, 0):
-        from nt import _getfinalpathname
-    else:
-        raise ImportError
-except (AttributeError, ImportError):
-    # On Windows XP and earlier, two files are the same if their absolute
-    # pathnames are the same.
-    # Non-Windows operating systems fake this method with an XP
-    # approximation.
-    def _getfinalpathname(f):
-        return normcase(abspath(f))
+    from nt import _getfinalpathname
+
+    # determine if two files are in fact the same file
+    def realpath(filename):
+        filename = os.fspath(filename)
+        extended_length_prefix = '\\\\?\\' if isinstance(filename, str) else b'\\\\?\\'
+        is_extended_path = filename.startswith(extended_length_prefix)
+        unresolved = filename if is_extended_path else abspath(filename)
+        resolved_parts = []
+        while True:
+            try:
+                resolved_parts.append(_getfinalpathname(unresolved))
+                break
+            except OSError:
+                unresolved, tail = split(unresolved)
+                if not tail:
+                    resolved_parts.append(unresolved)
+                    break
+                resolved_parts.append(tail)
+        resolved = join(*reversed(resolved_parts))
+        # try to convert extended path to normal if
+        # initial path did not use \\?\ prefix and result uses it
+        if not is_extended_path and resolved.startswith(extended_length_prefix):
+            resolved = _extended_to_normal(resolved)
+        return resolved
+
+    def _extended_to_normal(path):
+        drive, rest = splitdrive(path)
+        if not rest:
+            return path
+        if isinstance(path, str):
+            sep = '\\'
+            colon = ':'
+            letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+            extended_unc_prefix = '\\\\?\\UNC'
```

Comment:
FYI, "UNC" is natively "\Global??\UNC", a symbolic link to "\Device\Mup", the multiple UNC device. That's why I had called it `unc_dev`. Calling the UNC device path a prefix works for me. I just want to clarify that "UNC" is an actual object in the object-manager namespace; it's not a magic string prefix like "\\\\?\\" (which natively gets replaced by "\\??\\", another magic string that tells the object manager to use the fast-path local/global lookup using the cached device directories of the process).